### PR TITLE
Fix incorrect SFT loss logging during gradient accumulation

### PR DESCRIPTION
**Pull request by notkisk**
This PR fixes a logging bug in the **SFT training script** where `train_loss` only reflected the **last micro-batch** of each gradient accumulation cycle instead of the average across all micro-batches

The fix accumulates detached losses on-device and computes their mean after all accumulation steps  ensuring accurate loss reporting without affecting gradients or training behavior

---

### **Before**

```python
train_loss = loss.detach()  # only logs last micro-batch
```

### **After**

```python
accumulated_loss += loss.detach()
...
train_loss = accumulated_loss / grad_accum_steps
```

---

**Impact:**
✅ Correct average loss logging
✅ No change to gradient correctness or training dynamics

### CODE REVIEW

**Comment by notkisk** [comment]
In the original code:

for micro_step in range(grad_accum_steps):
    ...
    train_loss = loss.detach()  # for logging
    loss = loss / grad_accum_steps
    loss.backward()


The line

train_loss = loss.detach()


overwrites the previous value each micro-step - so after the loop, train_loss only contains the last micro-batch’s loss, not the average across all micro-batches.

That means the loss logging (the one printed and sent to WandB) was incorrect.
It did not represent the true per-step average loss, even though the gradients and training behavior were correct (since the loss was divided before backward()).

**Comment by karpathy** [comment]
It's not a bug, it's a deliberate choice. Doing "the right thing" adds a bunch of added bloat/complexity synchronizing the number. We're only seeing/logging the loss at a single rank but that's ok as just an estimate. The training is all correct otherwise.
